Puzzle.py

```diff
@@ -0,0 +1,269 @@
+        It is generally assumed that the puzzle is square, so rectangular puzzles will have rectangular pieces.
+
+    Inputs:
+
+        image:
+            A PIL image, returned by Image.open('some_image_file.png').
+
+        pieces_per_side:
+            Determines how many pieces the puzzle is cut into. 
+            will slightly shrink the image if the pieces cannot fit evenly into it.
+            ^^^ This will leave a border around the image
+
+        output_size:
+            This is the size of the image as displayed on the screen.
+
+        puzzle_pos:
+            this is used to determine which piece was clicked when
+            the puzzle is not in the top-left corner of the window.
+
+    Methods:
+
+        modify_image(image, output_size)
+            Resizes the input image to the output size.
+            Splits the image into total_pieces parts, then creates a PuzzlePiece object from the image.
+            PuzzlePiece object is stored in a list called pieces.
+            Raises an Exception if the resized image cannot be divided into total_pieces without remainder.
+
+        get_tile_index_from_pos(mouse_position):
+            Uses puzzle_pos and mouse_position to determine which puzzle piece the mouse is currently over.
+            Returns the orderlist index of the tile
+
+        get_tile_from_pos(mouse_position):
+            Calls get_tile_index_from_pos(mouse_position)
+            returns the pieces index of the tile, which can be used to access the tile through pieces[index]
+
+        generate_orderlist():
+            re-creates the orderlist from the relative_x and relative_y values of every PuzzlePiece
+            it is a good idea to call this before using image_update()
+
+        image_update():
+            updates the puzzle image to reflect the order in orderlist. This should be called
+            at the end of every loop where the puzzle is changed
+
+    Attributes:
+
+    pieces_per_side:
+        This is the number of pieces on any given section of the puzzle ( see below )
+        if pieces_per_side is 5:
+            puzzle looks like:
+                   r * * * * * 
+                   | * * * * * 
+           5 pieces| * * * * *
+                   | * * * * *
+                   L * * * * *
+        and total_pieces is 25.
+
+    image:
+        this is how the puzzle currently looks. On initialization, the puzzle looks like the input image.
+
+    orderlist:
+        This is a list of length total_pieces containing the order of the pieces.
+        For instance if pieces_per_side is 2, then orderlist starts as [0, 1, 2, 3]
+        [ 0 , 1 ,
+          2 , 3 ]
+        swapping the top two pieces:
+        orderlist is [1, 0, 2, 3]
+        puzzle looks like:
+        [ 1 , 0 ,
+          2 , 3 ]
+'''
+
+
+class Puzzle:
+    def __init__(self, image, pieces_per_side, output_size, puzzle_pos):
```

Comment:
I notice a lack of type hints! You can probably do:

```suggestion
    def __init__(self, image: PIL.Image, pieces_per_side: int, output_size: int, puzzle_pos: tuple[int, int]):
```

I think! Just make sure mypy is fine with that. I think you can probably figure out the methods' type hints!